bokehjs/test/integration/plots.ts

```diff
@@ -107,6 +107,101 @@ describe("Plot", () => {
     })
   })
 
+  describe("should support windowed auto-ranging", async () => {
+    function plot(window_axis: WindowAxis) {
+      const r = new Range1d({start: 0, end: 6})
+      const dr = new DataRange1d({range_padding: 0})
+      const p = figure({
+        sizing_mode: "fixed",
+        width: 300,
+        height: 300,
+        window_axis,
+        title: `window_axis == ${window_axis}`,
+        x_range: window_axis == "x" ? r: dr,
+        y_range: window_axis == "x" ? dr: r,
+      })
+      const s = p.scatter({x: [1, 2, 3, 4, 5], y: [1, 2, 3, 4, 5], size: 8})
+      return {p, s, r, dr}
+    }
+    for (const wax of ["x", "y"]) {
+      it(`with window_axis="${wax}" when data changes`, async () => {
+        const nwax = (wax == "x") ? "y" : "x"
+        const {p, s, dr} = plot(wax as WindowAxis)
+        const {view} = await display(p)
+
+        expect(dr.start).to.be.equal(1)
+        expect(dr.end).to.be.equal(5)
+
+        const data: any = {}
+
+        // updated data in fixed range
+        data[wax] = [1, 2, 3, 4, 5]
+        data[nwax] = [1, 2, 10, 4, 5]
+        s.data_source.data = data
+        await view.ready
+        expect(dr.start).to.be.equal(1)
+        expect(dr.end).to.be.equal(10)
+
+        // updated data not in fixed range (no change)
+        data[wax] = [-1, 1, 2, 3, 4, 100]
+        data[nwax] = [-100, 1, 2, 10, 4, 100]
+        s.data_source.data = data
+        await view.ready
+        expect(dr.start).to.be.equal(1)
+        expect(dr.end).to.be.equal(10)
+
+        // streamed data in fixed range
+        const new_data: any = {}
+        new_data[wax] = [2.5, 3.5]
+        new_data[nwax] = [-90, 100]
+        s.data_source.stream(new_data)
+        await view.ready
+        expect(dr.start).to.be.equal(-90)
+        expect(dr.end).to.be.equal(100)
+
+        // streamed data not in fixed range (no change)
+        new_data[wax] = [-1]
+        new_data[nwax] = [200]
+        s.data_source.stream(new_data)
+        await view.ready
+        expect(dr.start).to.be.equal(-90)
+        expect(dr.end).to.be.equal(100)
+
+        // Error: expected -90 to be equal to -10
+        data[wax] = [1, 2, 3, 4, 5]
+        data[nwax] = [1, 2, 10, 4, -10]
+        s.data_source.data = data
+        await view.ready
+        expect(dr.start).to.be.equal(-10)
+        expect(dr.end).to.be.equal(10)
```

Comment:
@mattpap I can't figure out why this one test is failing. The behavior is consistent as if the spatial index had not be completely rebuilt after a data update, but AFAIK it should be getting rebuilt. Do you see anything obviously wrong with the test or can you suggest how to debug this?

Reply:
Well, now its the first test above that is failing. I am at a loss, running examples all work perfectly as expected. I think there is some order of ops or race that can be an issue in the tests that I don't understand

Reply:
I was able to get all these passing by:

* `await`-ing twice in a row
* using an entirely new dict object for each block

I am not sure if this says something about the implementation, or some assumptions in the test machinery, or something else.

Reply:
`view.ready` promise isn't bullet proof, especially if long chains of updates, especially async updates, are involved. Worst case scenario, you can use `poll()` function from `core/util/defer`, to wait for a value to change. I use it sporadically if I can't get `view.ready` to work.

Reply:
sometimes the values in question are not supposed to change, so I just added a comment regarding the repeated calls. 

https://github.com/bokeh/bokeh/pull/14353/commits/dce2abb00664cc98a4001323f388716d35ebbbf4